server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -327,13 +338,57 @@ protected void setMinSeqNoToKeep(long seqNo) {
         this.minSeqNoToKeep = seqNo;
     }
 
-    @Override
-    void deleteReaderFiles(TranslogReader reader) {
+    private void deleteRemoteGeneration(Set<Long> generations) {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            translogTransferManager.deleteTranslogAsync(primaryTermSupplier.getAsLong(), generations);
+        } catch (IOException e) {
+            logger.error(() -> new ParameterizedMessage("Exception occurred while deleting generation {}", generations), e);
+        }
+    }
+
+    public void trimUnreferencedReaders() throws IOException {
+        // clean up local translog files and updates readers
+        super.trimUnreferencedReaders();
+
+        // cleans up remote translog files not referenced in latest uploaded metadata.
+        // This enables us to restore translog from the metadata in case of failover or relocation.
+        Set<Long> generationsToDelete = new HashSet<>();
+        for (long generation = minRemoteGenReferenced - 1; generation >= 0; generation--) {
+            if (fileTransferTracker.uploaded(Translog.getFilename(generation)) == false) {
+                break;
+            }
+            generationsToDelete.add(generation);
+        }
+        deleteRemoteGeneration(generationsToDelete);
+        deleteOlderPrimaryTranslogFilesFromRemoteStore();
+    }
+
+    private void deleteOlderPrimaryTranslogFilesFromRemoteStore() {
+        // The deletion of older translog files in remote store is on best-effort basis, there is a possibility that there
+        // are older files that are no longer needed and should be cleaned up. In here, we delete all files that are part
+        // of older primary term.
+        if (olderPrimaryCleaned.trySet(Boolean.TRUE)) {
+            logger.info("Cleaning up translog uploaded by previous primaries");
+            for (long oldPrimaryTerm = current.getPrimaryTerm() - 1; oldPrimaryTerm >= 0; oldPrimaryTerm--) {
```

Comment:
Shouldn't this initialized according to primary term mapped to min translog generation? If min translog generation is mapped to primary term X, oldPrimaryTerm should initialize to X - 1.

Also, do we need to go till 0 each time? This means time to execute this method will increase as primaryTerm increases. Can we identify if path does not exist as part of `onFailure` and break the loop?

Reply:
>Shouldn't this initialized according to primary term mapped to min translog generation? If min translog generation is mapped to primary term X, oldPrimaryTerm should initialize to X - 1.

Yes. Incorporated the comment.

>Also, do we need to go till 0 each time? This means time to execute this method will increase as primaryTerm increases. Can we identify if path does not exist as part of onFailure and break the loop?

No, made changes here too. Now, I am reading all primary terms that exists in the remote store and deleting those which are smaller than the minimum referenced primary term in the latest metadata file.